docs/proposals/api-types-package-structure.md

```diff
@@ -0,0 +1,475 @@
+│   ├── types.go
+│   └── types_swagger_doc_generated.go
+└── validation
+    ├── doc.go
+    ├── events.go
+    ├── events_test.go
+    ├── schema.go
+    ├── schema_test.go
+    ├── testdata
+    │   └── v1
+    │       ├── invalidPod1.json
+    │       ├── invalidPod2.json
+    │       ├── invalidPod3.json
+    │       ├── invalidPod.yaml
+    │       └── validPod.yaml
+    ├── validation.go
+    └── validation_test.go
+
+```
+
+The api in `pkg/apis` is closer to where we want to be.
+
+```
+pkg/apis/
+└── extensions
+    ├── deep_copy_generated.go
+    ├── helpers.go
+    ├── helpers_test.go
+    ├── install
+    │   ├── install.go
+    │   └── install_test.go
+    ├── register.go
+    ├── types.generated.go
+    ├── types.go
+    ├── v1beta1
+    │   ├── conversion_generated.go
+    │   ├── conversion.go
+    │   ├── deep_copy_generated.go
+    │   ├── defaults.go
+    │   ├── defaults_test.go
+    │   ├── register.go
+    │   ├── types.generated.go
+    │   ├── types.go
+    │   └── types_swagger_doc_generated.go
+    └── validation
+        ├── validation.go
+        └── validation_test.go
+```
+
+## Desired directory structure
+
+We want to define a standard layout and structure for Kubernetes API types.
+Below is a first draft, with explanation afterwards. We may try to move the
+relevant parts of the legacy `pkg/api/v1` api into the new structure, although
+this api doesn't have a group.
+
+```
+apis
+├── group
+│   ├── install
+│   │   └── install_all_versions.go
+│   ├── types.go
+│   ├── v1
+│   │   ├── conversion
+│   │   │   ├── generated_conversion.go
+│   │   │   └── manual_conversion_overrides.go
+│   │   ├── defaulting
+│   │   │   ├── generated_defaulters.go
+│   │   │   └── generated_default_stubs.go
+│   │   ├── install
+│   │   │   └── install_all_resources.go
+│   │   ├── types.go
+│   │   └── validation
+│   │       ├── generated_validation.go
+│   │       └── manual_validation_overrides.go
+│   ├── v1beta1
+│   │   ├── conversion
+│   │   │   ├── generated_conversion.go
+│   │   │   └── manual_conversion_overrides.go
+│   │   ├── defaulting
+│   │   │   ├── generated_defaulters.go
+│   │   │   └── generated_default_stubs.go
+│   │   ├── install
+│   │   │   └── install_all_resources.go
+│   │   ├── types.go
+│   │   └── validation
+│   │       ├── generated_validation.go
+│   │       └── manual_validation_overrides.go
+│   └── v2alpha1
+│       ├── conversion
+│       │   ├── generated_conversion.go
+│       │   └── manual_conversion_overrides.go
+│       ├── defaulting
+│       │   ├── generated_defaulters.go
+│       │   └── generated_default_stubs.go
+│       ├── install
+│       │   └── install_all_resources.go
+│       ├── types.go
+│       └── validation
+│           ├── generated_validation.go
+│           └── manual_validation_overrides.go
+└── unversioned
+    ├── meta
+    │   └── meta.go
+    ├── resource
+    │   └── quantity.go
+    ├── time
+    │   └── time.go
+    └── types.go
+```
```

Comment:
We now also have encoding and decoding code generation in the same package as types.go. This is only really necessary for the server or high-performance clients. It would be nice to make it optional as well.

Currently, it's implemented as methods on the types. To separate this, you'd have to type aliases in another package, generate the methods there, and have a serializer/deserializer smart enough to do the type conversions.